### ISSUE TITLE
RFC: Package resource management

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#6243]
In SilverStripe 3 client assets were typically referenced by their path relative to the `BASE_PATH`. This includes the module folder, which is typically hardcoded.

This has simple approach has served us for a number of years but it makes a number of things difficult:
- Moving the webroot to something other than the project root, which is a common security practise.
- Moving modules into the `vendor/` folder, which would make it easier to work with a larger number of modules, and therefore create smaller, more single-purpose modules.
- Making use of modules without carefully setting up a project tree. The "simpler test run" branches of framework and cms hit this, and overcame them by refining the use of constants like `CMS_DIR`. However it's awkward to scale this out to every module.

We'd like to improve on this.

## Recommendation

I recommend that we introduce a "resource locator" as an injected service. The service would be responsible for handling both server-accessed and client-accessed resources.

A mock-up of the relevant interfaces are here: https://gist.github.com/sminnee/92d5486ba9d23d12ef4253b5ef274bf7

The Injector would be responsible for providing one of these.

One key aspect of these interfaces is that modules are referenced by their full composer package names.

### Requirements

Right now we have a call such as:

``` php
Requirements::javascript(FRAMEWORK_ADMIN_DIR . '/client/dist/js/vendor.js');
```

This could be replaced with an inlined call to the ClientResourceLocator, but this is very verbose:

``` php
Requirements::javascript(
  Injector::inst()
    ->get('SilverStripe\\Core\\ResourceLocator')
    ->getResource('framework/admin', 'client/dist/js/vendor.js')
    ->getURL()
);
```

So my recommendation is that we have additional support methods in Requirements that handle this:

``` php
Requirements::packageJavascript('framework/admin', 'client/dist/js/vendor.js');
```

Or in a template:

```
<% require packageJavascript('framework/admin', 'client/dist/js/vendor.js') %>
```

### Finding modules

For finding the location of the packages, I would expect a relatively simple implementation such as this one would suffice by default. Potentially, package searching could also be a pluggable Injector service.

``` php
function pathFor($package) {
   if(file_exists(BASE_PATH . '/vendor/' . $package)) {
    return BASE_PATH . '/vendor/' . $package;
   }

  $shortname = substr($package, strpos($package)+1);
   if(file_exists(BASE_PATH . '/' . $shortname)) {
     return BASE_PATH . '/' . $shortname;
   }

   $packageJson = json_decode(file_get_contents(BASE_PATH . '/composer.json'));
   if(isset($packageJson['name']) && $packageJson['name'] == $package) {
     return BASE_PATH;
   }
}
```
### ResourceLocator implementation

There may end up being many implementations of ResourceLocator, but I can see 3 very useful approaches that may be part of core:
- **Naive resource locator:** assumes BASE_PATH is in the webroot and returns a URL relative to that. This would "just work" and may be the dfault
- **On-demand resource locator:** Would be configured with both a base path within with to place client resources, and the corresponding URL of that folder. For example, the path might be `/public_html/client` and the URL might be `/client/`. When a client asset is requested, it copies it in if it's missing or out of date and then returns its url. This would be good for dev environments.
- **Pre-cached resource locator:** Configured the same way as the on-demand resource locator, but it would have a build task that copies all the assets into the resource-path. If you don’t run that build task, assets won’t be available—no checking it done at load time. This would be good for production deployments.

Note that the server-side aspect of resource locator will be much simpler. In the original version of this API we had 2 separate interfaces, but this didn't work nicely with a 

More advanced approaches would include, for example, uploading client resources to S3, or running an npm-based build toolchain to prepare the resources.

### You have to include the current package name on your resource includes

In the status quo, you just refer to a client-side resource by name. It's not very flexible, but it is simple:

```
Requirements::javascript('client/dist/js/vendor.js')
```

Now you're going to need to put a composer package alongside every client resource. If you're building a module, that might be okay, but for a project this might seem a little strange:

```
Requirements::packageJavascript('sminnee/myproject', 'client/dist/js/vendor.js')
```

It's difficult to see how you could determine the owner of the caller package without any help, but perhaps by passing in `__FILE__` you could determine the current package based on that.

```
Requirements::forPackage('silverstripe/admin')->javascript('client/dist/bundle.js');
Requirements::forOwnerOf(__FILE__)->javascript('client/dist/bundle.js');
```

There are many ways that this API could be desiged, eg:

```
Requirements::javascript(Package::byName('silverstripe/framework')->getResource('client/dist/bundle.js'));
Requirements::javascript(Package::ownerOf(__FILE__)->getResource('client/dist/bundle.js'));
```
## Alternatives

The changes to the `Requirements` API could be given more thought. For example, a more fluent syntax may be better:

``` php
Requirements::inst()->withResource('framework/admin', 'client/dist/js/vendor.js')->insertScript();
Requirements::inst()->withResource('framework/admin', 'client/dist/stlyes/bundle.css')->insertStyleLink([ 'media' => 'screen' ]);
```

**Comment by tractorcow** [comment]
👍 looks good; My addition would be to include a `hasResource` method to query each module for resources.

I would also suggest a convention for referring to resources, similar to the convention for templates.

e.g.

```php
$path = 'mysite/javascript/script.js`; // "naive" resource
$path = [
	'module' => 'silverstripe/admin',
	'client/dist/script.js'
]; // module scoped resource.
```

Then you could consolidate the api.

```php
$url = Injector::inst()
    ->get('SilverStripe\\Core\\ClientResourceLocator')
    ->getResourceURL($path);
$fullPath = Injector::inst()
    ->get('SilverStripe\\Core\\ServerResourceLocator')
    ->getResourcePath($path);
```

Moving path / url resolution to as late in the chain as possible means that you can easily infer either client or server path at any point; We store the resource specification not the url / path.

**Comment by sminnee** [comment]
> My addition would be to include a hasResource method to query each module for resources.

Agreed, although I think we'd want a `hasClientResource()` as well, which would return false, e.g. for `some/file.php`

> I would also suggest a convention for referring to resources, similar to the convention for templates.

I don't think we want to have naive resources. At best, you might have resources from the current package, but I'm not sure how you'd check that. Naive resources would be like direct references to superglobals — they only serve to cause bugs. Furthermore a "naive resource" is the equivalent of not suing the resource locator at all, so if you really had a reason to do that you could simply just avoid this package.

You've suggested that a 2 element array is a better method signature than a pair of arguments, which I don't think is the case. However, since I think naive resources should never exist, maybe that's moot?

> Moving path / url resolution to as late in the chain as possible means that you can easily infer either client or server path at any point; We store the resource specification not the url / path. The use case in requirements is that we need to occasionally generate both local filesystem reference as well as public url, so we should be able to convert between the two.

If we needed this I would probably have a `getResource()` method that returns an object, rather than relying on a magic array format for this.

In practise, however, most of these resources references are directly linked to calls such as `Requirements::javascript()`

**Comment by tractorcow** [comment]
I'm fine with an object instead of an array for resource. :)

**Comment by tractorcow** [comment]
@chillu do you have a preference for how we store resource references?

**Comment by sminnee** [comment]
@tractorcow Can you point to an example in the current codebase where this intermediary resource object (or array) would be of benefit?

**Comment by tractorcow** [comment]
Requirements_Backend::pathForFile() is called when emitting all requirements to the browser.

This method would take the resource, and then:
 - Use the local path to validate the file exists
 - Return the public url

**Comment by sminnee** [comment]
Ok cool, so a 2-argument syntax for Requirements::javascript would internally store a Resource object

**Comment by sminnee** [comment]
Upon further thought, I think that this change overlaps greatly with ThemeResourceLoader. So I think that

 * The ClientResourceLocator should replace ThemeResourceLoader
 * The ClientResourceLocator should accept an array of package names as well as a single one

**Comment by chillu** [comment]
> One key aspect of these interfaces is that modules are referenced by their full composer package > Requirements::packageJavascript('framework/admin', 'client/dist/js/vendor.js');

This confused me a bit, since `framework/admin` isn't a composer package but rather a subfolder inside the `silverstripe/framework` package. Once we [move out](https://github.com/silverstripe/silverstripe-framework/issues/5641) the subfolder into its own package, the name will probably be `silverstripe/admin`. So unless I'm misunderstanding your API proposal, the example would be more consistent as:

> Requirements::packageJavascript('silverstripe/framework', 'admin/client/dist/js/vendor.js');

> The ClientResourceLocator should replace ThemeResourceLoader

Agree on the overlap with [ThemeResourceLoader](https://github.com/silverstripe/silverstripe-framework/blob/master/src/View/ThemeResourceLoader.php). So you see the first argument in `ClientResourceLocator` retaining the ability to specify a composite package and theme path like `vendor/mymodule:mytheme`?

Footnotes:
- We should extend the concept of referencing modules by their composer package name to other places in core as well, possibly building on [ClassManifest->getModules()](https://github.com/silverstripe/silverstripe-framework/blob/master/src/Core/Manifest/ClassManifest.php#L319). The [FileFinder](https://github.com/silverstripe/silverstripe-framework/blob/master/src/Assets/FileFinder.php) API (and specialisations like [ManifestFileFinder](https://github.com/silverstripe/silverstripe-framework/blob/master/src/Core/Manifest/ManifestFileFinder.php)) are separate concerns, they're just traversing folders which are expected to be modules (coming off `BASE_PATH`).  
- We'll eventually need metadata to indicate which module client resources to copy into a public webroot. Are we OK to continue hardcoding this path into `Requirements` calls? In the example above, `admin/client/dist` would be copied into the webroot, but there's not always a clear 1:1 mapping with a "client resource path", right?
- The proposed approach is similar to [Symfony Assetic](http://symfony.com/doc/3.1/assetic/asset_management.html#including-javascript-files), which uses the bundle name (rather than composer package name) as an interpolated string (e.g. `@AppBundle/Resources/public/js/myfile.js`). Composer package names seem superior to introducing a separate notion of a "package/bundle name".

**Comment by sminnee** [comment]
> This confused me a bit, since framework/admin isn't a composer package but rather a subfolder inside the silverstripe/framework package. Once we move out the subfolder into its own package, the name will probably be silverstripe/admin. So unless I'm misunderstanding your API proposal.

Sorry, I wrote it for the bold new world in which admin was moved to a separate composer package. ;-)

**Comment by chillu** [comment]
Also, stumbled on a [resource locator discussion](https://groups.google.com/forum/#!topic/php-fig/WMaKNNhHZJw) from 2013 over at PHP FIG. Only glanced over it, it advocates using namespaces (as the associated PSR lookup rules) to identify paths for non-PHP assets. I see the composer package name as superior, but thought it would be interesting to contrast approaches here.

**Comment by sminnee** [comment]
Yeah – it doesn't look like that idea got very far. If it was promoted to a more widely accepted RFC I might consider jumping on board for the sake of cross-project interoperability, but that approach would raise a lot of other questions and issues that I don't think are worth jumping into.

**Comment by sminnee** [comment]
I've implemented a much simpler solution in #7065 that covers most of what we need here. I think we can put this RFC to bed and work with what we've got in SS4.
